File: blocks/block_04_simulink_and_fixed_point/python/lab_4_2_fixed_point_digital_mixer.py
#!/usr/bin/env python3
"""Lab 4.2 — Fixed-point digital mixer.

This script generates a complex IQ tone, shifts it with a floating-point mixer,
implements an educational Q1.15 fixed-point mixer, estimates EVM/spurs and saves
IEEE-style figures.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np


@dataclass(frozen=True)
class LabConfig:
    fs: float = 2.4e6
    sample_count: int = 32768
    input_tone_hz: float = 120e3
    shift_hz: float = -120e3
    noise_rms: float = 0.01
    q_fractional_bits: int = 15
    phase_bits: int = 24
    seed: int = 11


ROOT = Path(__file__).resolve().parents[3]
ASSET_DIR = ROOT / "docs" / "assets"


def q15(v: np.ndarray, fractional_bits: int = 15) -> np.ndarray:
    scale = 2**fractional_bits
    return np.clip(np.round(v * scale), -32768, 32767).astype(np.int16)


def spectrum_db(x: np.ndarray, fs: float) -> tuple[np.ndarray, np.ndarray]:
    window = np.hanning(len(x))
    coherent_gain = np.sum(window) / len(window)
    xw = x * window
    spec = np.fft.fftshift(np.fft.fft(xw)) / (len(xw) * coherent_gain)
    freq = np.fft.fftshift(np.fft.fftfreq(len(xw), d=1 / fs))
    mag_db = 20 * np.log10(np.maximum(np.abs(spec), 1e-15))
    return freq, mag_db


def fixed_point_mixer_q15(x: np.ndarray, osc: np.ndarray, fractional_bits: int) -> tuple[np.ndarray, int]:
    scale = 2**fractional_bits
    xi = q15(np.real(x), fractional_bits)
    xq = q15(np.imag(x), fractional_bits)
    ci = q15(np.real(osc), fractional_bits)
    sq = q15(np.imag(osc), fractional_bits)

    n = len(x)
    yi = np.zeros(n, dtype=np.int16)
    yq = np.zeros(n, dtype=np.int16)
    saturation_count = 0

    for k in range(n):
        acc_i = int(xi[k]) * int(ci[k]) - int(xq[k]) * int(sq[k])
        acc_q = int(xi[k]) * int(sq[k]) + int(xq[k]) * int(ci[k])

        ri = int(np.round(acc_i / scale))
        rq = int(np.round(acc_q / scale))
        ri_sat = int(np.clip(ri, -32768, 32767))
        rq_sat = int(np.clip(rq, -32768, 32767))
        saturation_count += int(ri != ri_sat) + int(rq != rq_sat)

        yi[k] = ri_sat
        yq[k] = rq_sat

    y = (yi.astype(np.float64) + 1j * yq.astype(np.float64)) / scale
    return y, saturation_count


def estimate_peak_frequency(x: np.ndarray, fs: float) -> float:
    window = np.hanning(len(x))
    spec = np.fft.fftshift(np.fft.fft(x * window))
    freq = np.fft.fftshift(np.fft.fftfreq(len(x), d=1 / fs))
    return float(freq[int(np.argmax(np.abs(spec)))])


def estimate_spur_dbc(x: np.ndarray, fs: float, exclude_bins: int = 6) -> float:
    window = np.hanning(len(x))
    spec = np.abs(np.fft.fftshift(np.fft.fft(x * window)))
    peak_idx = int(np.argmax(spec))
    peak = float(spec[peak_idx])
    masked = spec.copy()
    lo = max(0, peak_idx - exclude_bins)
    hi = min(len(spec), peak_idx + exclude_bins + 1)
    masked[lo:hi] = 0.0
    spur = float(np.max(masked))
    return float(20 * np.log10(max(spur, 1e-15) / max(peak, 1e-15)))


def save_plot_spectrum(cfg: LabConfig, x: np.ndarray, y_float: np.ndarray, y_fixed: np.ndarray) -> None:
    ASSET_DIR.mkdir(parents=True, exist_ok=True)
    f, xdb = spectrum_db(x, cfg.fs)
    _, yfdb = spectrum_db(y_float, cfg.fs)
    _, yqdb = spectrum_db(y_fixed, cfg.fs)

    plt.figure(figsize=(7.0, 4.2))
    plt.plot(f / 1e3, xdb, label="input")
    plt.plot(f / 1e3, yfdb, label="float mixer")
    plt.plot(f / 1e3, yqdb, label="fixed mixer")
    plt.grid(True, alpha=0.35)
    plt.xlabel("Frequency, kHz")
    plt.ylabel("Magnitude, dBFS")
    plt.title("Lab 4.2 — Fixed-point mixer spectrum comparison")
    plt.legend(loc="best")
    plt.tight_layout()
    plt.savefig(ASSET_DIR / "lab42_fixed_point_mixer_spectrum.png", dpi=180)
    plt.close()


def save_plot_error(cfg: LabConfig, err: np.ndarray) -> None:
    f, edb = spectrum_db(err, cfg.fs)

    plt.figure(figsize=(7.0, 4.2))
    plt.plot(f / 1e3, edb)
    plt.grid(True, alpha=0.35)
    plt.xlabel("Frequency, kHz")
    plt.ylabel("Error magnitude, dBFS")
    plt.title("Lab 4.2 — Fixed-point mixer error spectrum")
    plt.tight_layout()
    plt.savefig(ASSET_DIR / "lab42_fixed_point_mixer_error.png", dpi=180)
    plt.close()


def save_plot_phase_resolution(cfg: LabConfig) -> None:
    bits = np.array([12, 16, 20, 24, 28, 32])
    resolution = cfg.fs / (2.0**bits)

    plt.figure(figsize=(7.0, 4.2))
    plt.semilogy(bits, resolution, marker="o")
    plt.grid(True, which="both", alpha=0.35)
    plt.xlabel("Phase accumulator width, bits")
    plt.ylabel("Frequency resolution, Hz")
    plt.title("Lab 4.2 — NCO frequency resolution")
    plt.tight_layout()
    plt.savefig(ASSET_DIR / "lab42_nco_frequency_resolution.png", dpi=180)
    plt.close()


def main() -> None:
    cfg = LabConfig()
    rng = np.random.default_rng(cfg.seed)
    t = np.arange(cfg.sample_count) / cfg.fs

    x = np.exp(1j * 2 * np.pi * cfg.input_tone_hz * t)
    x += cfg.noise_rms * (rng.standard_normal(cfg.sample_count) + 1j * rng.standard_normal(cfg.sample_count))
    x = x / np.max(np.abs(x)) * 0.80

    phase_increment = int(round(cfg.shift_hz / cfg.fs * (2**cfg.phase_bits)))
    actual_shift_hz = phase_increment * cfg.fs / (2**cfg.phase_bits)
    phase = 2 * np.pi * (np.arange(cfg.sample_count) * phase_increment % (2**cfg.phase_bits)) / (2**cfg.phase_bits)
    osc = np.exp(1j * phase)

    y_float = x * np.exp(1j * 2 * np.pi * cfg.shift_hz * t)
    y_fixed, saturation_count = fixed_point_mixer_q15(x, osc, cfg.q_fractional_bits)

    err = y_float - y_fixed
    rms_error = float(np.sqrt(np.mean(np.abs(err) ** 2)))
    signal_rms = float(np.sqrt(np.mean(np.abs(y_float) ** 2)))
    evm_pct = float(100 * rms_error / max(signal_rms, 1e-15))
    expected_output_hz = cfg.input_tone_hz + cfg.shift_hz
    measured_peak_hz = estimate_peak_frequency(y_fixed, cfg.fs)
    frequency_shift_error_hz = measured_peak_hz - expected_output_hz
    spur_dbc = estimate_spur_dbc(y_fixed, cfg.fs)
    delta_f = cfg.fs / (2**cfg.phase_bits)

    save_plot_spectrum(cfg, x, y_float, y_fixed)
    save_plot_error(cfg, err)
    save_plot_phase_resolution(cfg)

    print("Lab 4.2 — Fixed-point digital mixer")
    print(f"Input tone: {cfg.input_tone_hz/1e3:.1f} kHz")
    print(f"Requested shift: {cfg.shift_hz/1e3:.1f} kHz")
    print(f"Actual NCO shift: {actual_shift_hz/1e3:.6f} kHz")
    print(f"Phase bits: {cfg.phase_bits}")
    print(f"NCO frequency resolution: {delta_f:.6f} Hz")
    print(f"Measured output peak: {measured_peak_hz:.3f} Hz")
    print(f"Frequency shift error: {frequency_shift_error_hz:.3f} Hz")
    print(f"RMS error: {rms_error:.6e}")
    print(f"EVM: {evm_pct:.4f} %")
    print(f"Largest spur estimate: {spur_dbc:.2f} dBc")
    print(f"Saturation count: {saturation_count}")
    print(f"Figures saved to: {ASSET_DIR}")


if __name__ == "__main__":
    main()

File: blocks/block_04_simulink_and_fixed_point/matlab/lab_4_2_fixed_point_digital_mixer.m
% Lab 4.2 — Fixed-point digital mixer
% Generates a complex IQ tone, shifts it with float and Q1.15 fixed-point
% mixers, estimates EVM/spurs and saves IEEE-style figures.

clear; close all; clc;

rootDir = fullfile(fileparts(mfilename('fullpath')), '..', '..', '..');
assetDir = fullfile(rootDir, 'docs', 'assets');
if ~exist(assetDir, 'dir')
    mkdir(assetDir);
end

rng(11);
fs = 2.4e6;
N = 32768;
inputToneHz = 120e3;
shiftHz = -120e3;
noiseRms = 0.01;
fracBits = 15;
phaseBits = 24;
scale = 2^fracBits;

t = (0:N-1).' / fs;
x = exp(1j*2*pi*inputToneHz*t) + noiseRms*(randn(N,1) + 1j*randn(N,1));
x = x ./ max(abs(x)) * 0.80;

phaseIncrement = round(shiftHz / fs * 2^phaseBits);
actualShiftHz = phaseIncrement * fs / 2^phaseBits;
phase = 2*pi*mod((0:N-1).' * phaseIncrement, 2^phaseBits) / 2^phaseBits;
osc = exp(1j*phase);

yFloat = x .* exp(1j*2*pi*shiftHz*t);

xi = int16(clip(round(real(x) * scale), -32768, 32767));
xq = int16(clip(round(imag(x) * scale), -32768, 32767));
ci = int16(clip(round(real(osc) * scale), -32768, 32767));
sq = int16(clip(round(imag(osc) * scale), -32768, 32767));

yi = zeros(N, 1, 'int16');
yq = zeros(N, 1, 'int16');
saturationCount = 0;

for k = 1:N
    accI = int64(xi(k)) * int64(ci(k)) - int64(xq(k)) * int64(sq(k));
    accQ = int64(xi(k)) * int64(sq(k)) + int64(xq(k)) * int64(ci(k));
    ri = round(double(accI) / scale);
    rq = round(double(accQ) / scale);
    riSat = clip(ri, -32768, 32767);
    rqSat = clip(rq, -32768, 32767);
    saturationCount = saturationCount + double(ri ~= riSat) + double(rq ~= rqSat);
    yi(k) = int16(riSat);
    yq(k) = int16(rqSat);
end

yFixed = (double(yi) + 1j*double(yq)) / scale;
err = yFloat - yFixed;
rmsError = sqrt(mean(abs(err).^2));
signalRms = sqrt(mean(abs(yFloat).^2));
evmPct = 100 * rmsError / max(signalRms, eps);
expectedOutputHz = inputToneHz + shiftHz;
measuredPeakHz = estimatePeakFrequency(yFixed, fs);
frequencyShiftErrorHz = measuredPeakHz - expectedOutputHz;
spurDbc = estimateSpurDbc(yFixed);
deltaF = fs / 2^phaseBits;

[f, xDb] = spectrumDb(x, fs);
[~, yFloatDb] = spectrumDb(yFloat, fs);
[~, yFixedDb] = spectrumDb(yFixed, fs);
figure('Color', 'w');
plot(f/1e3, xDb, 'DisplayName', 'input'); hold on;
plot(f/1e3, yFloatDb, 'DisplayName', 'float mixer');
plot(f/1e3, yFixedDb, 'DisplayName', 'fixed mixer');
grid on;
xlabel('Frequency, kHz'); ylabel('Magnitude, dBFS');
title('Lab 4.2 — Fixed-point mixer spectrum comparison');
legend('Location', 'best');
saveas(gcf, fullfile(assetDir, 'lab42_fixed_point_mixer_spectrum_matlab.png'));

[fErr, errDb] = spectrumDb(err, fs);
figure('Color', 'w');
plot(fErr/1e3, errDb);
grid on;
xlabel('Frequency, kHz'); ylabel('Error magnitude, dBFS');
title('Lab 4.2 — Fixed-point mixer error spectrum');
saveas(gcf, fullfile(assetDir, 'lab42_fixed_point_mixer_error_matlab.png'));

bits = [12 16 20 24 28 32];
resolution = fs ./ (2.^bits);
figure('Color', 'w');
semilogy(bits, resolution, '-o');
grid on;
xlabel('Phase accumulator width, bits'); ylabel('Frequency resolution, Hz');
title('Lab 4.2 — NCO frequency resolution');
saveas(gcf, fullfile(assetDir, 'lab42_nco_frequency_resolution_matlab.png'));

fprintf('Lab 4.2 — Fixed-point digital mixer\n');
fprintf('Input tone: %.1f kHz\n', inputToneHz/1e3);
fprintf('Requested shift: %.1f kHz\n', shiftHz/1e3);
fprintf('Actual NCO shift: %.6f kHz\n', actualShiftHz/1e3);
fprintf('Phase bits: %d\n', phaseBits);
fprintf('NCO frequency resolution: %.6f Hz\n', deltaF);
fprintf('Measured output peak: %.3f Hz\n', measuredPeakHz);
fprintf('Frequency shift error: %.3f Hz\n', frequencyShiftErrorHz);
fprintf('RMS error: %.6e\n', rmsError);
fprintf('EVM: %.4f %%\n', evmPct);
fprintf('Largest spur estimate: %.2f dBc\n', spurDbc);
fprintf('Saturation count: %d\n', saturationCount);
fprintf('Figures saved to: %s\n', assetDir);

function y = clip(x, lo, hi)
    y = min(max(x, lo), hi);
end

function [freq, magDb] = spectrumDb(x, fs)
    n = length(x);
    w = hann(n);
    coherentGain = sum(w) / n;
    spec = fftshift(fft(x .* w)) / (n * coherentGain);
    freq = fftshift((-floor(n/2):ceil(n/2)-1).' * fs / n);
    magDb = 20*log10(max(abs(spec), 1e-15));
end

function peakHz = estimatePeakFrequency(x, fs)
    n = length(x);
    w = hann(n);
    spec = fftshift(abs(fft(x .* w)));
    freq = fftshift((-floor(n/2):ceil(n/2)-1).' * fs / n);
    [~, idx] = max(spec);
    peakHz = freq(idx);
end

function spurDbc = estimateSpurDbc(x)
    n = length(x);
    w = hann(n);
    spec = fftshift(abs(fft(x .* w)));
    [peak, peakIdx] = max(spec);
    excludeBins = 6;
    lo = max(1, peakIdx - excludeBins);
    hi = min(n, peakIdx + excludeBins);
    spec(lo:hi) = 0;
    spur = max(spec);
    spurDbc = 20*log10(max(spur, 1e-15) / max(peak, 1e-15));
end
